File: frontend/mobile/lib/services/homework.service.dart
import 'package:dio/dio.dart';

import '../common/HomeworkTimeFormat.dart';
import '../interfaces/homework.interface.dart';
import 'api.service.dart';

class HomeworkException implements Exception {
  final String message;

  const HomeworkException(this.message);

  @override
  String toString() => message;
}

abstract class HomeworkRepository {
  Future<List<HomeworkTaskData>> getPendingHomework();
  Future<List<HomeworkSubject>> getSubjects();
  Future<HomeworkTaskData> createHomework(CreateHomeworkInput input);
  Future<HomeworkTaskData> updateHomework(
    HomeworkTaskData task,
    UpdateHomeworkInput input,
  );
  Future<void> deleteHomework(int workloadId);
  Future<void> finishHomework(int workloadId);
}

class HomeworkService implements HomeworkRepository {
  final ApiService _apiService;

  HomeworkService({ApiService? apiService})
    : _apiService = apiService ?? ApiService();

  @override
  Future<List<HomeworkTaskData>> getPendingHomework() async {
    try {
      final response = await _apiService.get('/user/workload/pending');
      final body = Map<String, dynamic>.from(response.data as Map);
      final data = body['data'];
      if (data is! List) return const [];
      return data
          .whereType<Map>()
          .map(
            (item) =>
                HomeworkTaskData.fromJson(Map<String, dynamic>.from(item)),
          )
          .toList()
        ..sort((left, right) => left.deadline.compareTo(right.deadline));
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถโหลดรายการงานได้');
    } on TypeError {
      throw const HomeworkException('ข้อมูลรายการงานจากระบบไม่ถูกต้อง');
    }
  }

  @override
  Future<HomeworkTaskData> updateHomework(
    HomeworkTaskData task,
    UpdateHomeworkInput input,
  ) async {
    try {
      await _apiService.put(
        '/user/workload/update/${task.workloadId}',
        data: {
          'workload_name': input.assignment.trim(),
          'deadline_date': _dateValue(input.deadline),
          'deadline_time': formatHomeworkApiTime(input.deadline),
          'note': input.note.trim().isEmpty ? null : input.note.trim(),
        },
      );
      return task.copyWith(
        assignment: input.assignment.trim(),
        deadline: input.deadline,
        note: input.note.trim(),
      );
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถแก้ไขงานได้');
    }
  }

  @override
  Future<List<HomeworkSubject>> getSubjects() async {
    try {
      final response = await _apiService.get(
        '/user/workload/subjects',
        queryParameters: {'schedule_type_id': 1},
      );
      final body = Map<String, dynamic>.from(response.data as Map);
      final data = body['data'];
      if (data is! List) return const [];
      return data
          .whereType<Map>()
          .map(
            (item) => HomeworkSubject.fromJson(Map<String, dynamic>.from(item)),
          )
          .toList();
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถโหลดรายวิชาได้');
    } on TypeError {
      throw const HomeworkException('ข้อมูลรายวิชาจากระบบไม่ถูกต้อง');
    }
  }

  @override
  Future<void> deleteHomework(int workloadId) async {
    try {
      await _apiService.delete('/user/workload/delete/$workloadId');
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถลบงานได้');
    }
  }

  @override
  Future<HomeworkTaskData> createHomework(CreateHomeworkInput input) async {
    try {
      final response = await _apiService.post(
        '/user/workload/add',
        data: {
          'schedule_time_id': input.subject.scheduleTimeId,
          'workload_type_id': input.type.id,
          'workload_name': input.assignment,
          'deadline_date': _dateValue(input.deadline),
          'deadline_time': formatHomeworkApiTime(input.deadline),
          'note': input.note.trim().isEmpty ? null : input.note.trim(),
        },
      );
      final body = Map<String, dynamic>.from(response.data as Map);
      return HomeworkTaskData(
        workloadId: _asInt(body['workload_id']),
        scheduleTimeId: input.subject.scheduleTimeId,
        workloadTypeId: input.type.id,
        workloadTypeName: '${body['workload_type_name'] ?? input.type.name}',
        subjectId: input.subject.subjectId,
        subject: input.subject.subjectName,
        assignment: input.assignment,
        deadline: input.deadline,
        dueDate:
            '${input.deadline.day}/${input.deadline.month}/${input.deadline.year + 543}',
        dueTime: formatHomeworkDisplayTime(input.deadline),
        note: input.note,
      );
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถเพิ่มงานได้');
    } on TypeError {
      throw const HomeworkException('ข้อมูลการเพิ่มงานจากระบบไม่ถูกต้อง');
    }
  }

  @override
  Future<void> finishHomework(int workloadId) async {
    try {
      await _apiService.put('/user/workload/finish/$workloadId');
    } on DioException catch (error) {
      throw _toHomeworkException(error, 'ไม่สามารถส่งงานได้');
    }
  }

  HomeworkException _toHomeworkException(DioException error, String fallback) {
    final data = error.response?.data;
    if (data is Map && data['message'] is String) {
      return HomeworkException(data['message'] as String);
    }
    return HomeworkException(fallback);
  }
}

String _dateValue(DateTime value) =>
    '${value.year.toString().padLeft(4, '0')}-'
    '${value.month.toString().padLeft(2, '0')}-'
    '${value.day.toString().padLeft(2, '0')}';

int _asInt(dynamic value) {
  if (value is int) return value;
  return int.tryParse('$value') ?? 0;
}

File: frontend/web/user/src/services/homework.service.ts
import axios, { AxiosInstance } from "axios";
import Cookies from "js-cookie";
import type {
  CreateHomeworkInput,
  HomeworkSubject,
  HomeworkTask,
  UpdateHomeworkInput,
} from "@/interfaces/homework.interface";

interface HomeworkApiRow {
  workload_id: number | string;
  schedule_time_id: number | string;
  workload_type_id: number | string;
  workload_type_name?: string;
  subject_id?: string;
  subject_name?: string;
  workload_name?: string;
  deadline_date?: string;
  deadline_time?: string;
  note?: string | null;
}

const pad = (value: number) => value.toString().padStart(2, "0");

const dateValue = (value: Date) =>
  `${value.getFullYear()}-${pad(value.getMonth() + 1)}-${pad(value.getDate())}`;

const timeValue = (value: Date) =>
  `${pad(value.getHours())}:${pad(value.getMinutes())}:${pad(value.getSeconds())}`;

const parseDeadline = (date: string | undefined, time: string | undefined) => {
  const datePart = (date ?? "").slice(0, 10);
  const timePart = (time ?? "00:00:00").slice(0, 8);
  const parsed = new Date(`${datePart}T${timePart}`);
  return Number.isNaN(parsed.getTime()) ? new Date() : parsed;
};

const toTask = (row: HomeworkApiRow): HomeworkTask => ({
  workload_id: Number(row.workload_id),
  schedule_time_id: Number(row.schedule_time_id),
  workload_type_id: Number(row.workload_type_id),
  workload_type_name: row.workload_type_name ?? "งาน",
  subject_id: row.subject_id ?? "",
  subject_name: row.subject_name ?? "",
  workload_name: row.workload_name ?? "",
  deadline: parseDeadline(row.deadline_date, row.deadline_time),
  note: row.note ?? "",
});

class HomeworkService {
  private apiClient: AxiosInstance;

  constructor() {
    this.apiClient = axios.create({
      baseURL: process.env.NEXT_PUBLIC_API_URL || "http://localhost:4000",
      timeout: 10000,
    });
    this.apiClient.interceptors.request.use((config) => {
      const accessToken = Cookies.get("accessToken");
      if (accessToken) config.headers.Authorization = `Bearer ${accessToken}`;
      return config;
    });
  }

  async getPendingHomework(): Promise<HomeworkTask[]> {
    try {
      const response = await this.apiClient.get<{ data?: HomeworkApiRow[] }>(
        "/user/workload/pending"
      );
      return (response.data.data ?? []).map(toTask).sort(
        (left, right) => left.deadline.getTime() - right.deadline.getTime()
      );
    } catch (error) {
      throw this.toError(error, "ไม่สามารถโหลดรายการงานได้");
    }
  }

  async getSubjects(): Promise<HomeworkSubject[]> {
    try {
      const response = await this.apiClient.get<{ data?: HomeworkSubject[] }>(
        "/user/workload/subjects",
        { params: { schedule_type_id: 1 } }
      );
      return response.data.data ?? [];
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) return [];
      throw this.toError(error, "ไม่สามารถโหลดรายวิชาได้");
    }
  }

  async createHomework(
    input: CreateHomeworkInput,
    subjects: HomeworkSubject[]
  ): Promise<HomeworkTask> {
    try {
      const response = await this.apiClient.post<HomeworkApiRow>(
        "/user/workload/add",
        {
          schedule_time_id: input.schedule_time_id,
          workload_type_id: input.workload_type_id,
          workload_name: input.workload_name.trim(),
          deadline_date: dateValue(input.deadline),
          deadline_time: timeValue(input.deadline),
          note: input.note.trim() || null,
        }
      );
      const subject = subjects.find(
        (item) => item.schedule_time_id === input.schedule_time_id
      );
      return {
        workload_id: Number(response.data.workload_id),
        schedule_time_id: input.schedule_time_id,
        workload_type_id: input.workload_type_id,
        workload_type_name: response.data.workload_type_name ?? "งาน",
        subject_id: subject?.subject_id ?? "",
        subject_name: subject?.subject_name ?? "",
        workload_name: input.workload_name.trim(),
        deadline: input.deadline,
        note: input.note.trim(),
      };
    } catch (error) {
      throw this.toError(error, "ไม่สามารถเพิ่มงานได้");
    }
  }

  async updateHomework(
    task: HomeworkTask,
    input: UpdateHomeworkInput
  ): Promise<HomeworkTask> {
    try {
      await this.apiClient.put(`/user/workload/update/${task.workload_id}`, {
        workload_name: input.workload_name.trim(),
        deadline_date: dateValue(input.deadline),
        deadline_time: timeValue(input.deadline),
        note: input.note.trim() || null,
      });
      return {
        ...task,
        workload_name: input.workload_name.trim(),
        deadline: input.deadline,
        note: input.note.trim(),
      };
    } catch (error) {
      throw this.toError(error, "ไม่สามารถแก้ไขงานได้");
    }
  }

  async deleteHomework(workloadId: number): Promise<void> {
    try {
      await this.apiClient.delete(`/user/workload/delete/${workloadId}`);
    } catch (error) {
      throw this.toError(error, "ไม่สามารถลบงานได้");
    }
  }

  async finishHomework(workloadId: number): Promise<void> {
    try {
      await this.apiClient.put(`/user/workload/finish/${workloadId}`);
    } catch (error) {
      throw this.toError(error, "ไม่สามารถส่งงานได้");
    }
  }

  private toError(error: unknown, fallback: string): Error {
    if (axios.isAxiosError(error)) {
      return new Error(error.response?.data?.message || fallback);
    }
    return error instanceof Error ? error : new Error(fallback);
  }
}

export const homeworkService = new HomeworkService();
export default homeworkService;
